### ISSUE TITLE
Provide compatibility with previous Keras-Preprocessing API.

### ISSUE THREAD

**Issue by Dref360** [tensorflow/tensorflow#40203]
**System information**
- TensorFlow version (you are using): 2.2
- Are you willing to contribute it (Yes/No): Yes

In response to https://github.com/keras-team/keras-preprocessing/issues/299

**Describe the feature and the current behavior/state.**

Currently, matching image datasets iterations with real pathnames is not as trivial as it was in keras-preprocessing. It is also not explained clearly in the documentation.

```python
# Keras preprocessing

flow = ImageDataGenerator.flow_from_directory('a_path', ...)
filenames = flow.filenames

# TF Dataset 

dataset = `image_dataset_from_directory('a_path', ...)
image_paths, labels, class_names = keras.preprocessing.dataset_utils.index_directory('a_path')
```

**Suggestion**
Include a Dataset that will return filenames as an attribute of the returned Dataset.

This dataset would also be shuffled, batched so that if we zip both datasets, we can easily track what is the filename currently being used. This is useful when debugging models, tracking performance, etc.

**Will this change the current api? How?**

No, we add an attribute to the returned Dataset.

**Who will benefit with this feature?**

Every user using `flow_from_directory(...).filenames`.

**Any Other info.**

I can submit a PR, but I wanted to know if this is a good fit or should we suggest users use the example above.

**Comment by L-Eriksson** [comment]
Thank you @Dref360. I think my intention has been included in your submission clearly and precisely, so nothing to add. I personally would vote for the solution just proposed instead of the example above.

**Comment by paultanger** [comment]
Hi, I am interested in this and wondering if there is a way I could contribute to this?  It seems like a simple solution could be add an optional parameter to the image_dataset_from_directory function for those users that want the image paths, such as: 

``` python
tf.keras.preprocessing.image_dataset_from_directory(
    directory,
    labels="inferred",
    label_mode="int",
    class_names=None,
    color_mode="rgb",
    batch_size=32,
    image_size=(256, 256),
    shuffle=True,
    seed=None,
    validation_split=None,
    subset=None,
    interpolation="bilinear",
    follow_links=False,
    ## new parameter ##
    img_paths=False
)
```

Then at the end of the function, just:

``` python
if img_paths:
   return dataset, image_paths
```
This shouldn't break anything else.  Alternatively, I like the dataset attribute approach as well (it seems a bit more OOP but also maybe not applicable to datasets that are not image datasets?).  Is anyone working on either solution?  I would be happy to help contribute to it.

**Comment by tanzhenyu** [comment]
Yep, this should be fairly simple, contributions welcome!

**Comment by paultanger** [comment]
I can submit a pull request, wasn't sure exactly how to document it, would that be better to discuss in the pull request?  Otherwise, what about adding this to the docstring:

`img_paths: Whether to return the filepaths of the images read from the dir.
`

```
If img_paths=True, it also returns a list of filepaths for each image in
    the dataset.
```

**Comment by tanzhenyu** [comment]
> I can submit a pull request, wasn't sure exactly how to document it, would that be better to discuss in the pull request? Otherwise, what about adding this to the docstring:
> 
> `img_paths: Whether to return the filepaths of the images read from the dir. `
> 
> ```
> If img_paths=True, it also returns a list of filepaths for each image in
>     the dataset.
> ```

I think you can turn on by default, without any API change? This doesn't seem like a backward-incompatible change to me.

**Comment by paultanger** [comment]
> > I can submit a pull request, wasn't sure exactly how to document it, would that be better to discuss in the pull request? Otherwise, what about adding this to the docstring:
> > `img_paths: Whether to return the filepaths of the images read from the dir. `
> > ```
> > If img_paths=True, it also returns a list of filepaths for each image in
> >     the dataset.
> > ```
> 
> I think you can turn on by default, without any API change? This doesn't seem like a backward-incompatible change to me.

If people have scripts setup only expecting a dataset to be returned it could break things, no?  It would put the dataset and image paths into a tuple.

**Comment by tanzhenyu** [comment]
> > > I can submit a pull request, wasn't sure exactly how to document it, would that be better to discuss in the pull request? Otherwise, what about adding this to the docstring:
> > > `img_paths: Whether to return the filepaths of the images read from the dir. `
> > > ```
> > > If img_paths=True, it also returns a list of filepaths for each image in
> > >     the dataset.
> > > ```
> > 
> > 
> > I think you can turn on by default, without any API change? This doesn't seem like a backward-incompatible change to me.
> 
> If people have scripts setup only expecting a dataset to be returned it could break things, no? It would put the dataset and image paths into a tuple.

You just need to add a `dataset.file_names = ...`, right?

**Comment by paultanger** [comment]
Currently it just creates a Dataset not TFRecordDataset so it doesn't have a file_names attribute, but could change that.. maybe more involved.  Would it make sense to do that inside this paths_and_labels_to_dataset? 

``` python
def paths_and_labels_to_dataset(image_paths,
                                image_size,
                                num_channels,
                                labels,
                                label_mode,
                                num_classes,
                                interpolation):
  """Constructs a dataset of images and labels."""
  # TODO(fchollet): consider making num_parallel_calls settable
  path_ds = dataset_ops.Dataset.from_tensor_slices(image_paths)
  img_ds = path_ds.map(
      lambda x: path_to_image(x, image_size, num_channels, interpolation))
  if label_mode:
    label_ds = dataset_utils.labels_to_dataset(labels, label_mode, num_classes)
    img_ds = dataset_ops.Dataset.zip((img_ds, label_ds))
  return img_ds
````

**Comment by tanzhenyu** [comment]
> Currently it just creates a Dataset not TFRecordDataset so it doesn't have a file_names attribute, but could change that.. maybe more involved. Would it make sense to do that inside this paths_and_labels_to_dataset?
> 
> ```python
> def paths_and_labels_to_dataset(image_paths,
>                                 image_size,
>                                 num_channels,
>                                 labels,
>                                 label_mode,
>                                 num_classes,
>                                 interpolation):
>   """Constructs a dataset of images and labels."""
>   # TODO(fchollet): consider making num_parallel_calls settable
>   path_ds = dataset_ops.Dataset.from_tensor_slices(image_paths)
>   img_ds = path_ds.map(
>       lambda x: path_to_image(x, image_size, num_channels, interpolation))
>   if label_mode:
>     label_ds = dataset_utils.labels_to_dataset(labels, label_mode, num_classes)
>     img_ds = dataset_ops.Dataset.zip((img_ds, label_ds))
>   return img_ds
> ```

Yes the assumption here is that once you get the Dataset object, you can retrieve `file_names` from it otherwise the information will get lost after map or batch functions, or any conversion to TFRecordDataset in your case.

**Comment by paultanger** [comment]
Hmm, not exactly sure where to do that (I'm thinking this: `dataset_ops.Dataset.file_paths = image_paths`), and if it needs to be shuffled along with the batches?  Also, I notice that the way class_names is defined is just an instance attribute that then gets lost after batch functions etc.. so maybe whatever we do with file_names can also be done with class_names to be carried through the pipeline and accessible later?

**Comment by tanzhenyu** [comment]
> Hmm, not exactly sure where to do that (I'm thinking this: `dataset_ops.Dataset.file_paths = image_paths`), and if it needs to be shuffled along with the batches? Also, I notice that the way class_names is defined is just an instance attribute that then gets lost after batch functions etc.. so maybe whatever we do with file_names can also be done with class_names to be carried through the pipeline and accessible later?

Yeah exactly what I meant above -- you can define them as instance attribute, which will get lost but the hope would be you retrieve this information before any further processing. That's probably the only way to work with `Dataset` today
